gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java

```diff
@@ -0,0 +1,246 @@
+        convertedOptions[index++] = option.convert(blobInfo);
+      }
+      return convertedOptions;
+    }
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided {@code BlobInfo}. The storage service is used
+   * to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param info blob's info
+   */
+  public Blob(Storage storage, BlobInfo info) {
+    this.storage = checkNotNull(storage);
+    this.info = checkNotNull(info);
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided bucket and blob names. The storage service is
+   * used to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param bucket bucket's name
+   * @param blob blob's name
+   */
+  public Blob(Storage storage, String bucket, String blob) {
+    this.storage = checkNotNull(storage);
+    this.info = BlobInfo.of(bucket, blob);
+  }
+
+  /**
+   * Get the blobs's information.
+   * 
+   * @return a {@code BlobInfo} object for this blob
+   */
+  public BlobInfo info() {
+    return info;
+  }
+
+  /**
+   * Check if this blob exists.
+   *
+   * @return true if this blob exists, false otherwise
+   * @throws StorageException upon failure
+   */
+  public boolean exists() {
+    return storage.get(info.bucket(), info.name()) != null;
+  }
+
+  /**
+   * Return this blob's content.
+   *
+   * @param options blob read options
+   * @return the blob's content
+   * @throws StorageException upon failure
+   */
+  public byte[] content(Storage.BlobSourceOption... options) {
+    return storage.readAllBytes(info.bucket(), info.name(), options);
+  }
+
+  /**
+   * Update the blob's information. Bucket or blob's name cannot be changed by this method. If you
+   * want to rename the blob or move it to a different bucket use the {@link #copyTo} and
+   * {@link #delete} operations.
+   *
+   * @param blobInfo new blob's information. Bucket and blob names must match the current ones
+   * @param options update options
+   * @throws StorageException upon failure
+   */
+  public void update(BlobInfo blobInfo, BlobTargetOption... options) {
+    checkArgument(Objects.equals(blobInfo.bucket(), info.bucket()), "Bucket name must match");
+    checkArgument(Objects.equals(blobInfo.name(), info.name()), "Blob name must match");
+    info = storage.update(blobInfo, options);
+  }
+
+  /**
+   * Delete this blob.
+   *
+   * @param options blob delete options
+   * @return true if blob was deleted
+   * @throws StorageException upon failure
+   */
+  public boolean delete(BlobSourceOption... options) {
+    return storage.delete(info.bucket(), info.name(), convert(info, options));
+  }
+
+  /**
+   * Copy this blob.
+   *
+   * @param target target blob
+   * @param options source blob options
+   * @return the copied blob
+   * @throws StorageException upon failure
+   */
+  public Blob copyTo(BlobInfo target, BlobSourceOption... options) {
```

Comment:
As for **gcloud-go** I assume you mean "our `Blob.copy`"?
Looks like most libraries takes destination as a string.. Do they copy their own (this) metadata
to the destination?
 
> This This method does not update remote object's metadata ...

But it does, using the values from target `BlobInfo`, just not from the source.
We can iterate on the Javadoc to make it more clear however I think it is easier
to add later than remove later.

I agree, having this functionality in the other libraries makes it more compelling to have it
here as well, but maybe we can start with something simpler such as.
copyTo(bucket, sourceOptions...) and copyTo(bucket, name, sourceOptions....)
and copy the metadata by default?

> Why do you thing that the second version needs to be simplified? 

I consider the options as an advanced feature that most user will not care about and
would rather make it as transparent as possible.  Also, a side note, I am not sure the param
sequence feels right (target, sourceOptions, targetOptions)..